# Model Training

For predicting digits out of the MNIST dataset I am training a Convolutional Neural Network containing 1 convolutional followed by max pooling, dense and dropout layers.
As optimizer [ADAM](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam) is used while the loss is calculated based on [Sparse Categorical Crossentropy](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/sparse_categorical_crossentropy).

In [3]:
import tensorflow as tf

from data.file import model_convolutional_dataset
from data.load import load_train_dataset, load_test_dataset
from model.convolutional import build_compile_fit_dataset

In [4]:
batch_size = 32

x_train, y_train = load_train_dataset(normalize=True)
x_test, y_test = load_test_dataset(normalize=True)

train_input = tf.data.Dataset.from_tensor_slices((x_train, y_train)).batch(batch_size).prefetch(tf.data.AUTOTUNE)
test_input = tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(batch_size).prefetch(tf.data.AUTOTUNE)

In [5]:
model, history = build_compile_fit_dataset(train_input, optimizer='adam', loss='sparse_categorical_crossentropy',
                                           metrics=['accuracy'], epochs=10)
model.summary()

Epoch 1/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.2114 - accuracy: 0.9370
Epoch 2/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0856 - accuracy: 0.9740
Epoch 3/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0611 - accuracy: 0.9805
Epoch 4/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0451 - accuracy: 0.9851
Epoch 5/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0367 - accuracy: 0.9875
Epoch 6/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0290 - accuracy: 0.9902
Epoch 7/10
1875/1875 [==============================] - 16s 8ms/step - loss: 0.0249 - accuracy: 0.9920
Epoch 8/10
1875/1875 [==============================] - 16s 8ms/step - loss: 0.0211 - accuracy: 0.9927
Epoch 9/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.0201 - accuracy: 0.9931
Epoch 10/10
1875/1875 [==============================] - 16s 8ms/step - l

## Persist Model
The model was persisted under [data](../../data/model/convolutional_dataset) after training in order to be available for subsequent steps.

In [6]:
model.save(model_convolutional_dataset)

INFO:tensorflow:Assets written to: /Users/raffael/learn/fhnw/xai/xai-adversarial-attack/data/model/convolutional_dataset/assets


## Evaluation

In [9]:
model.evaluate(x_test, y_test, batch_size=10)

1000/1000 [==============================] - 2s 2ms/step - loss: 0.0673 - accuracy: 0.9839


[0.06729170680046082, 0.9839000105857849]

On the test dataset the model provides a performance of around **98%**.